core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java

```diff
@@ -0,0 +1,273 @@
+        ShardRouting shard = clusterState.routingTable().index("idx").shard(0).primaryShard();
+        RoutingAllocation routingAllocation =  newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, Collections.singleton(noRebalanceDecider)), clusterState);
+        RebalanceDecision rebalanceDecision = allocator.decideRebalance(shard, routingAllocation);
+        assertNotEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertEquals("rebalancing is not allowed", rebalanceDecision.getFinalExplanation());
+        assertNotNull(rebalanceDecision.getNodeDecisions());
+        assertNull(rebalanceDecision.getAssignedNodeId());
+        assertEquals(1, rebalanceDecision.getCanRebalanceDecision().getDecisions().size());
+        for (Decision subDecision : rebalanceDecision.getCanRebalanceDecision().getDecisions()) {
+            assertEquals("foobar", ((Decision.Single) subDecision).getExplanation());
+        }
+
+        assertAssignedNodeRemainsSame(allocator, routingAllocation, shard);
+    }
+
+    public void testRebalancePossible() {
+        AllocationDecider canAllocateDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                return Decision.YES;
+            }
+        };
+        Tuple<ClusterState, RebalanceDecision> rebalance = setupStateAndRebalance(canAllocateDecider, Settings.EMPTY, true);
+        ClusterState clusterState = rebalance.v1();
+        RebalanceDecision rebalanceDecision = rebalance.v2();
+        assertEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.YES, rebalanceDecision.getFinalDecisionType());
+        assertNotNull(rebalanceDecision.getFinalExplanation());
+        assertEquals(clusterState.nodes().getSize() - 1, rebalanceDecision.getNodeDecisions().size());
+        assertNotNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testRebalancingNotAllowedDueToCanAllocate() {
+        AllocationDecider canAllocateDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                return Decision.NO;
+            }
+        };
+        Tuple<ClusterState, RebalanceDecision> rebalance = setupStateAndRebalance(canAllocateDecider, Settings.EMPTY, false);
+        ClusterState clusterState = rebalance.v1();
+        RebalanceDecision rebalanceDecision = rebalance.v2();
+        assertEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertNotNull(rebalanceDecision.getFinalExplanation());
```

Comment:
Maybe there's a better explanation we can assert rather than just checking it has one?